shared/middleware/helpers/storage_apps.rb

```diff
@@ -140,24 +140,44 @@ def get_abuse(channel_id)
 
   # Determine if the current user can view the project
   def get_sharing_disabled(channel_id, current_user_id)
-    user_storage_owner_id, _ = storage_decrypt_channel_id(channel_id)
-    owner_id = user_storage_ids_table.where(id: user_storage_owner_id).first[:user_id]
+    owner_storage_id, storage_app_id = storage_decrypt_channel_id(channel_id)
+    owner_user_id = user_storage_ids_table.where(id: owner_storage_id).first[:user_id]
 
     # Sharing of a project is not disabled for the project owner
     # or the teachers of the project owner
-    if current_user_id == owner_id
+    # or if the current user paired with the owner
+    if current_user_id == owner_user_id
       return false
-    elsif teaches_student?(owner_id, current_user_id)
+    elsif teaches_student?(owner_user_id, current_user_id)
       return false
+    elsif get_user_sharing_disabled(owner_user_id)
+      !users_paired_on_level?(storage_app_id, current_user_id, owner_user_id, owner_storage_id)
     else
-      return get_user_sharing_disabled(owner_id)
+      return false
     end
 
   # Default to sharing disabled if there is an error
   rescue ArgumentError, OpenSSL::Cipher::CipherError
     true
   end
 
+  def users_paired_on_level?(storage_app_id, current_user_id, owner_user_id, owner_storage_id)
+    channel_tokens_table = DASHBOARD_DB[:channel_tokens]
+    level_id_row = channel_tokens_table.where(storage_app_id: storage_app_id, storage_id: owner_storage_id).first
+    return false if level_id_row.nil?
+    level_id = level_id_row[:level_id]
+
+    user_levels_table = DASHBOARD_DB[:user_levels]
+    owner_user_level_id = user_levels_table.select(:id).where(user_id: owner_user_id, level_id: level_id)
+    current_user_level_id = user_levels_table.select(:id).where(user_id: current_user_id, level_id: level_id)
+
+    paired_user_levels_table = DASHBOARD_DB[:paired_user_levels]
+    paired_level_row = paired_user_levels_table.where(driver_user_level_id: owner_user_level_id, navigator_user_level_id: current_user_level_id).first
+    return false if paired_level_row.nil?
```

Comment:
thanks for reworking this method. I know I said we'd look at optimization "later", but now I'm thinking that it is going to the be best and most practical to optimize this method by combining it into a single DB query using several joins (or maybe subqueries) now. Here is one example of some joins using Sequel: https://github.com/code-dot-org/code-dot-org/blob/4c2cdf25f3a0dd3faa5a96c9854271c841af1406/shared/middleware/helpers/auth_helpers.rb#L73-L82

here is another: https://github.com/code-dot-org/code-dot-org/blob/4c2cdf25f3a0dd3faa5a96c9854271c841af1406/dashboard/lib/projects_list.rb#L111-L128

here are some docs for doing joins in Sequel: https://github.com/jeremyevans/sequel/blob/master/doc/querying.rdoc#joins

Lastly, it may be easier to additionally `join` against the `user_storage_ids` table to convert between owner_user_id and owner_storage_id in the way that's compatible with SQL/Sequel joins, or it's possible you can discover some other way to do it.

I'd be happy to help you with this, now or in January :-)

Reply:
Actually, subqueries may be a better way to go than joins: http://sequel.jeremyevans.net/rdoc/files/doc/dataset_filtering_rdoc.html#label-Filtering+using+sub-queries 

interestingly, this may actually lead to an implementation which is very close to what you had originally implemented 😳 apologies for shooting first and asking questions later with all the comments!

Reply:
After offline discussion, it might actually be better to return early here rather than do an additional query. It shouldn't matter too much, given that not many students have sharing disabled.